### PR TITLE
feat(txnames): Write rules to project option

### PR DESCRIPTION
**Pull request by jjbayer**
To have a persistent store of rules discovered by the transaction clusterer, now also write rules to project options.

We still keep the redis store around, and merge newly discovered rules into both, such that we can later use this synchronized redis store to update rule lifetimes efficiently: The load of writes on every sanitized transaction name is very high, but the load of writes on the generation of new rules is low. Both postgres and redis can handle the latter, but only redis can handle the former.

See https://github.com/getsentry/team-ingest/issues/23

### CODE REVIEW

**Review by iker-barriocanal — APPROVED** [review]
LGTM. See the inheritance suggestion and your call to address the other suggestions on the comments. Feel free to merge it without requesting another review.

**Review by olksdr — APPROVED** [review]

File: src/sentry/ingest/transaction_clusterer/rules.py
Comment (jjbayer) [reply src/sentry/ingest/transaction_clusterer/rules.py:13]: Can't find a definitive answer to this question, although [here's](https://jellis18.github.io/post/2022-01-11-abc-vs-protocol/) a good writeup. I like how `Protocol` allows you to use [duck typing](https://en.wikipedia.org/wiki/Duck_typing) while still enforcing strict type checks.

File: src/sentry/ingest/transaction_clusterer/rules.py
Comment (jjbayer) [reply src/sentry/ingest/transaction_clusterer/rules.py:65]: Good point. If it's an issue in memory then we probably also should not write to project options, so maybe a general limit of ~ 200 rules per project would be a good idea.

File: src/sentry/ingest/transaction_clusterer/rules.py
Comment (jjbayer) [inline src/sentry/ingest/transaction_clusterer/rules.py:100]:
```suggestion
    """ Public interface for fetching rules for a project.
    Project options are the persistent, long-term store for rules, while redis is just a short-term buffer,
    so project options is what we fetch from. """
    return ProjectOptionRuleStore().read(project)
```

File: src/sentry/ingest/transaction_clusterer/rules.py
Comment (jjbayer) [reply src/sentry/ingest/transaction_clusterer/rules.py:110]: It actually contains the rules discovered by the latest clusterer run, which is a source of rules just like redis or postgres.

**Pull request merged**
